Refactor generated CPU XML

nvgoldin opened the pull request:
Docs PR: https://github.com/lago-project/lago/pull/481

Adding/changing the following options to ```LagoInitFile```:
#### Default ( no ``cpu_model`` or ``cpu`` configuration in the init file)
In the absence of any CPU configuration in the init file the generated XML will be:
```xml
...
<cpu mode="host-passthrough">
    <topology cores="1" sockets="2" threads="1"/>
</cpu>
<vcpu>2</vcpu>
...
```

### ``cpu_model: <model>`` 
LagoInitFile:
```yaml
...
    cpu_model: Westmere
    vcpu: 3
....
````
Generated XML:
```xml
<cpu match="exact">
    <model>Westmere</model>
    <topology cores="1" sockets="3" threads="1"/>
    <feature name="vmx" policy="require"/>
</cpu>
<vcpu>3</vcpu>
```
Note 1: it will attempt to add 'vmx' on Intel and 'svm' on AMD, if those are not available, it will issue a warning in the log.
Note 2: defining both ``cpu`` and ``cpu_model`` will throw an error.


### ``cpu_custom: ...``  - Custom CPU model
This is the interesting part which allows direct mapping from YAML to XML, the syntax is somewhat clumsy, but allows to test new and custom options directly from the init file, for example:
```yaml
domains:
  vm-el73:
    vcpu: 2
    cpu_custom:
      '@mode': custom
      '@match': exact
      model:
        '@fallback': allow
        '#text': Westmere
      feature:
        '@policy': optional
        '@name': 'vmx'
      numa:
        cell:
          -
            '@id': 0
            '@cpus': 0
            '@memory': 2048
            '@unit': 'MiB'
          -
            '@id': 1
            '@cpus': 1
            '@memory': 2048
            '@unit': 'MiB'
    memory: 4096
....
```
This will be the generated XML:
```xml
....
<cpu mode="custom" match="exact">
    <model fallback="allow">Westmere</model>
    <feature policy="optional" name="vmx"/>
    <numa>
        <cell id="0" cpus="0" memory="2048" unit="MiB"/>
        <cell id="1" cpus="1" memory="2048" unit="MiB"/>
    </numa>
    <topology cores="1" sockets="2" threads="1"/>
</cpu>
<vcpu>2</vcpu>
....
```
The conversion is pretty straight-forward, '@' maps to attributes of the key, and '#text' to its text. No validation is done, however if ``vcpu`` is defined, and ``topology`` is not, it will add it with sockets number equals to ``vcpu``.
I think using custom 'cpu' would be most useful to test new features and then enhance the YAML definition to create abbreviations for more specific CPU modes. For example, one possible abbreviation could be an ``auto-numa`` mode which will try to optimize(perhaps globally) the vcpus between numa nodes. 

#### ovirtlago changes
``get_ovirt_cpu_family`` will now report the matching CPU family of the first VDSM host it finds. If no VDSM host is available, it will report the one of the engine(a not so interesting case). If a different CPU model is needed for the cluster, it can be changed by defining the CPU model in the vdsm host init file. 

#### Mechanics
1. The first 3 commits are mainly code movement and little re-factoring: 
 -  libvirt-provider code is moved to ``lago/providers/libvirt/``
 - network definitions which resided in ``virt.py``, moved to ``lago/providers/libvirt/network.py``. 
2. The fourth commit adds the ``lago/providers/libvirt/cpu.py`` and updates ``lago/providers/libvirt/vm.py`` to fetch the XML from there.
3. The last commit adds initial unit testing, covering the basic cases for now.


#### TO-DO
- [x] Updated ovirtlago to the new model: instead of getting

- [x] Change ``'cpu:`` -> ``'cpu_custom'``
- [x] add unit tests
- [x] Do more OST tests against the new model. 

Less urgent:
1. Integrate some custom CPU configurations in Lago functional tests(without creating new suites, just use the existing ones).
2. Test how this behaves with a JSON formatted init file. Which also raises the question if we should keep supporting the JSON format.

nvgoldin commented:
ci test please

mz-pdm commented:
I tested the default configuration on my RHEL and in Fedora and CentOS virtual machines and it runs fine for me. I didn't test custom CPU configurations but the idea makes sense for me.

nvgoldin commented on lago/providers/libvirt/cpu.py:
```diff
@@ -0,0 +1,235 @@
+        flag = nested[vendor]
+        if self.host_cpu.find('feature/[@name="{0}"]'.format(flag)
+                              ) is not None:
+            cpu.append(self.generate_feature(name=flag))
```
reminder to my self:
This fails when emulating AMD on Intel(as it will require svm, but check vmx exists). Need to also compare the cpu_model vendor given by the user.

nvgoldin replied:
Fixed.

nvgoldin commented:
@mz-pdm  - Thanks for testing. 

Updated the PR with current progress.

nvgoldin commented:
ci test please

nvgoldin commented:
ci test please

nvgoldin commented:
ci test please

nvgoldin commented:
Once v0.36 is out I plan to merge this so we'll have enough testing time on unstable(if anyone still wants to review..)
OST basic-suite-master: http://jenkins.ovirt.org/job/ngoldin-ovirt-system-tests-manual-custom_lago/1/console

OST HE 4.0: http://jenkins.ovirt.org/job/ngoldin-ovirt-system-tests-manual-custom_lago/3/console

Unfortunately HE master manual jobs seems broken ATM.

nvgoldin commented:
Split out the oVirt logic to https://github.com/lago-project/lago/pull/470/commits/9d357c7f3865bffe453c8f3e4926c0bbd14c4582:
1. The mapping is defined in a yaml file ```data/ovirt_cpu_map.yaml```
2. It can be overriden by setting it in the init file under metadata: ``ovirt-cpu-map: path/to/file.yaml``
3. Also added ``IvyBridge -> SandyBridge`` mapping, following the logic presented in https://bugzilla.redhat.com/show_bug.cgi?id=1206701.

nvgoldin commented:
Verified again on basic-suite-master and HE 4.1.

nvgoldin commented:
ci merge please

nvgoldin commented:
ci merge please

nvgoldin commented:
ci merge please
